internal/notification/smtp_notifier.go

```diff
@@ -3,50 +3,106 @@ package notification
+	address  string
 }
 
 // NewSMTPNotifier create an SMTPNotifier targeting a given address.
 func NewSMTPNotifier(configuration schema.SMTPNotifierConfiguration) *SMTPNotifier {
 	return &SMTPNotifier{
+		username: configuration.Username,
+		password: configuration.Password,
+		sender:   configuration.Sender,
 		host:     configuration.Host,
 		port:     configuration.Port,
+		secure:   configuration.Secure,
 		address:  fmt.Sprintf("%s:%d", configuration.Host, configuration.Port),
-		sender:   configuration.Sender,
-		username: configuration.Username,
-		password: configuration.Password,
 	}
 }
 
-func (n *SMTPNotifier) authenticatedSend(recipient string, msg string) error {
-	auth := smtp.PlainAuth("", n.username, n.password, n.host)
-	err := smtp.SendMail(fmt.Sprintf("%s:%d", n.host, n.port), auth, n.sender,
-		[]string{recipient}, []byte(msg))
-	if err != nil {
-		return err
-	}
-	return nil
-}
+// Send send a identity verification link to a user.
+func (n *SMTPNotifier) Send(recipient string, subject string, body string) error {
+	msg := "From: " + n.sender + "\n" +
+		"To: " + recipient + "\n" +
+		"Subject: " + subject + "\n" +
+		"Content-Type: text/html\n" +
+		"MIME-version: 1.0;\nContent-Type: text/html; charset=\"UTF-8\";\n\n" +
+		body
 
-func (n *SMTPNotifier) unauthenticatedSend(recipient string, msg string) error {
-	// Connect to the remote SMTP server.
 	c, err := smtp.Dial(n.address)
 
 	if err != nil {
 		return err
 	}
 
+	//Do StartTLS if available (some servers only provide the auth extnesion after, and encrpytion is preferred)
+	starttls, _ := c.Extension("STARTTLS")
+	if starttls {
+		log.Debugf("SMTP server supports STARTTLS (verify cert: %t), attempting", n.secure)
+		tlsconfig := &tls.Config {
+			InsecureSkipVerify: !n.secure,
+			ServerName: n.host,
+		}
+		err := c.StartTLS(tlsconfig)
+		if err != nil {
+			return err
+		} else {
+			log.Debug("SMTP STARTTLS completed without error")
+		}
+	} else {
+		log.Debug("SMTP server does not support STARTTLS, skipping")
+	}
+
+	//Authenticate if possible
+	if n.password != "" {
+		authExtension, m := c.Extension("AUTH")
+		if authExtension {
+			log.Debugf("Config has SMTP password and server supports AUTH with the following mechanisms: %s", m)
+			mechanisms := strings.Split(m, " ")
+			var auth smtp.Auth
+
+			if starttls && utils.IsStringInSlice("PLAIN", mechanisms) {
+				auth = smtp.PlainAuth("", n.username, n.password, n.host)
+				log.Debug("SMTP server supports AUTH PLAIN, attempting")
+			} else if utils.IsStringInSlice("LOGIN", mechanisms) {
+				auth = utils.LoginAuth(n.username, n.password)
+				log.Debug("SMTP server supports AUTH LOGIN, attempting")
+			} else if utils.IsStringInSlice("CRAM-MD5", mechanisms) {
+				auth = smtp.CRAMMD5Auth(n.username, n.password)
+				log.Debug("SMTP server supports AUTH CRAM-MD5, attempting")
+			}
+			if auth == nil {
+				return errors.New("SMTP server does not advertise a AUTH mechanism that Authelia Supports")
```

Comment:
Idea for the future: we should probably introduce a hook at Authelia startup to perform early checks of configuration: "is SMTP server correctly configured?", "is db correctly configured?", "are credentials against service X ok?"
This way users could notice failures early during set up.

Reply:
This is very doable, and would probably force me to get more familiar with the code base.

Reply:
I will look into this. As you said future I think for the purposes of the PR you intend to resolve it prior to merging; but I'll leave this open for you to decide.

This part will probably involve two things, implement each stage of the send mail process into a function which returns an error or nil (i.e. STARTTLS, AUTH). I'm going to quickly see if I can implement this split of functionality now without breaking anything.

Reply:
I had a look at this, but I think it may require a bit more thought and full implementation instead of just splitting the functionality.

Reply:
Definitely not a quick win, this is why I mentioned "for the future" meaning later in another review. The PR as it is is fine for me, the goal is reached. Thanks.